File: services/messaging-agent/internal/middleware/circuit_breaker.go
package middleware

import (
	"sync"
	"sync/atomic"
	"time"
)

// CircuitState represents the circuit breaker FSM state.
type CircuitState int32

const (
	StateClosed   CircuitState = 0
	StateOpen     CircuitState = 1
	StateHalfOpen CircuitState = 2
)

// CircuitBreaker is a per-provider circuit breaker.
type CircuitBreaker struct {
	state            int32 // atomic CircuitState
	failures         int32
	failureThreshold int32
	recoveryTimeout  time.Duration
	lastFailure      time.Time
	mu               sync.Mutex
}

func NewCircuitBreaker(failureThreshold int, recoverySeconds int) *CircuitBreaker {
	return &CircuitBreaker{
		failureThreshold: int32(failureThreshold),
		recoveryTimeout:  time.Duration(recoverySeconds) * time.Second,
	}
}

// IsOpen returns true if the circuit is open (calls should be blocked).
func (cb *CircuitBreaker) IsOpen() bool {
	state := CircuitState(atomic.LoadInt32(&cb.state))
	if state == StateOpen {
		cb.mu.Lock()
		elapsed := time.Since(cb.lastFailure)
		cb.mu.Unlock()
		if elapsed > cb.recoveryTimeout {
			atomic.StoreInt32(&cb.state, int32(StateHalfOpen))
			return false
		}
		return true
	}
	return false
}

// RecordSuccess resets the circuit to Closed state.
func (cb *CircuitBreaker) RecordSuccess() {
	atomic.StoreInt32(&cb.failures, 0)
	atomic.StoreInt32(&cb.state, int32(StateClosed))
}

// RecordFailure increments failure count and opens the circuit if threshold reached.
func (cb *CircuitBreaker) RecordFailure() {
	count := atomic.AddInt32(&cb.failures, 1)
	if count >= cb.failureThreshold {
		cb.mu.Lock()
		cb.lastFailure = time.Now()
		cb.mu.Unlock()
		atomic.StoreInt32(&cb.state, int32(StateOpen))
	}
}

File: services/confidence-agent/src/middleware/circuit_breaker.rs
use std::sync::atomic::{AtomicU32, Ordering};
use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::sync::Mutex;

#[derive(Debug, Clone, PartialEq)]
pub enum CircuitState { Closed, Open, HalfOpen }

pub struct CircuitBreaker {
    state: Arc<Mutex<CircuitState>>,
    failure_count: Arc<AtomicU32>,
    failure_threshold: u32,
    recovery_timeout: Duration,
    last_failure: Arc<Mutex<Option<Instant>>>,
}

impl CircuitBreaker {
    pub fn new(failure_threshold: u32, recovery_timeout_secs: u64) -> Self {
        Self {
            state: Arc::new(Mutex::new(CircuitState::Closed)),
            failure_count: Arc::new(AtomicU32::new(0)),
            failure_threshold,
            recovery_timeout: Duration::from_secs(recovery_timeout_secs),
            last_failure: Arc::new(Mutex::new(None)),
        }
    }

    pub async fn is_open(&self) -> bool {
        let mut state = self.state.lock().await;
        if *state == CircuitState::Open {
            let last = self.last_failure.lock().await;
            if let Some(t) = *last {
                if t.elapsed() > self.recovery_timeout {
                    *state = CircuitState::HalfOpen;
                    return false;
                }
            }
            return true;
        }
        false
    }

    pub async fn record_failure(&self) {
        let count = self.failure_count.fetch_add(1, Ordering::SeqCst) + 1;
        if count >= self.failure_threshold {
            *self.state.lock().await = CircuitState::Open;
            *self.last_failure.lock().await = Some(Instant::now());
        }
    }

    pub async fn record_success(&self) {
        self.failure_count.store(0, Ordering::SeqCst);
        *self.state.lock().await = CircuitState::Closed;
    }
}
